io-engine/src/store/etcd.rs

```diff
@@ -55,17 +55,41 @@ impl Store for Etcd {
     async fn txn_kv<K: StoreKey>(
         &mut self,
         key: &K,
-        cmps: Vec<Compare>,
-        ops_success: Vec<TxnOp>,
-        ops_failure: Option<Vec<TxnOp>>,
-    ) -> Result<TxnResponse, StoreError> {
-        let fops = ops_failure.map_or(vec![], |v| v);
-        self.0
-            .txn(Txn::new().when(cmps).and_then(ops_success).or_else(fops))
+        new_value: Vec<u8>,
+        expected_value: Vec<u8>,
+    ) -> Result<Option<Vec<u8>>, StoreError> {
+        let cmp = Compare::value(key.to_string(), CompareOp::Equal, expected_value);
+        let put_op = TxnOp::put(key.to_string(), new_value, None);
+        let or_else_op = TxnOp::get(key.to_string(), None);
+
+        let txn_resp = self
+            .0
+            .txn(
+                Txn::new()
+                    .when([cmp])
+                    .and_then([put_op])
+                    .or_else([or_else_op]),
+            )
             .await
             .context(TxnErr {
                 key: key.to_string(),
-            })
+            })?;
+
+        if !txn_resp.succeeded() {
+            if let TxnOpResponse::Get(g) = &txn_resp.op_responses()[0] {
+                if let Some(kv) = g.kvs().first() {
+                    let current_value = kv.value();
+                    return Ok(Some(current_value.to_owned()));
+                }
+            } else {
+                // This is needed because absence of above response data shouldn't get
+                // implicitly assumed as success. With this the top level caller can
+                // compare this returned blank value with some expected value.
+                return Ok(Some(vec![]));
```

Comment:
Shouldn't we return a StoreError here as this seems like an etcd bug if we hit this?

Reply:
but won't vec![] lead into a program crash?

`let val = serde_json::from_slice::<NexusInfo>(&current_value).unwrap();`